libsplinter/src/error/internal.rs

```diff
@@ -0,0 +1,281 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std;
+    /// use splinter::error::InternalError;
+    ///
+    /// let io_err = std::io::Error::new(std::io::ErrorKind::Other, "io error");
+    /// let internal_error = InternalError::from_source(Box::new(io_err));
+    /// assert_eq!(format!("{}", internal_error), "io error");
+    /// ```
+    pub fn from_source(source: Box<dyn error::Error>) -> Self {
+        Self {
+            message: None,
+            source: Some(Source {
+                prefix: None,
+                source,
+            }),
+        }
+    }
+
+    /// Constructs a new `InternalError` from a specified source error and message string.
+    ///
+    /// The implementation of `std::fmt::Display` for this error will be the message string
+    /// provided.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std;
+    /// use splinter::error::InternalError;
+    ///
+    /// let io_err = std::io::Error::new(std::io::ErrorKind::Other, "io error");
+    /// let internal_error = InternalError::from_source_with_message(Box::new(io_err), "oops".to_string());
+    /// assert_eq!(format!("{}", internal_error), "oops");
+    /// ```
+    pub fn from_source_with_message(source: Box<dyn error::Error>, message: String) -> Self {
+        Self {
+            message: Some(message),
+            source: Some(Source {
+                prefix: None,
+                source,
+            }),
+        }
+    }
+
+    /// Constructs a new `InternalError` from a specified source error and prefix string.
+    ///
+    /// The implementation of `std::fmt::Display` for this error will be constructed from the
+    /// prefix and source message's display following the format of `format!("{}: {}", prefix,
+    /// source)`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std;
+    /// use splinter::error::InternalError;
+    ///
+    /// let io_err = std::io::Error::new(std::io::ErrorKind::Other, "io error");
+    /// let internal_error = InternalError::from_source_with_prefix(Box::new(io_err), "Could not open file".to_string());
+    /// assert_eq!(format!("{}", internal_error), "Could not open file: io error");
+    /// ```
+    pub fn from_source_with_prefix(source: Box<dyn error::Error>, prefix: String) -> Self {
+        Self {
+            message: None,
+            source: Some(Source {
+                prefix: Some(prefix),
+                source,
+            }),
+        }
+    }
+
+    /// Constructs a new `InternalError` with a specified message string.
+    ///
+    /// The implementation of `std::fmt::Display` for this error will be the message string
+    /// provided.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use splinter::error::InternalError;
+    ///
+    /// let internal_error = InternalError::with_message("oops".to_string());
+    /// assert_eq!(format!("{}", internal_error), "oops");
+    /// ```
+    pub fn with_message(message: String) -> Self {
```

Comment:
Is this meant to be different from the `from_` variant constructor methods?

Reply:
This is intentional from_source and with_message/with_prefix. In these terms "You create it from a source" and "You create it with a message" - and so when mixed, you end up with from_source_with_message, but individually from_source and with_message.

One reason against from_message is that in order to be consistent, it feels like when combined with source it should be from_source_and_message or from_message_and_source -- but it doesn't feel great to me.

Another reason is that an error is naturally generated from a source error. I.e. it is what the error is about. Message and prefix are closer to attributes being used rather than being converted/wrapped.

Reply:
That's a very good differentiator